service/dap/server.go

```diff
@@ -2959,6 +3037,74 @@ func (s *Server) doRunCommand(command string, asyncSetupDone chan struct{}) {
 	}
 }
 
+func (s *Server) run(command string, asyncSetupDone chan struct{}) (*api.DebuggerState, error) {
+	s.setRunning(true)
+	defer s.setRunning(false)
+
+	runningStep := command != api.Continue
+	if state, _ := s.debugger.State(true); state != nil {
+		// If state.NextInProgress, then we are in the middle of a step,
+		// regardless of the command.
+		runningStep = runningStep || state.NextInProgress
+	}
+
+	var state *api.DebuggerState
+	var err error
+	for {
+		state, err = s.resume(command, asyncSetupDone)
+		if state == nil || state.CurrentThread == nil || err != nil {
+			break
+		}
+
+		istp := s.handleLogPoints(state)
+		// Only resume execution if continue was the command used to run the program.
+		// Otherwise, the program will go past the step, next, stepout requests.
+		// TODO(suzmue): have s.debugger.Command() not clear internal breakpoints when
+		// hitting another breakpoint.
+		if s.debugger.StopReason() == proc.StopBreakpoint {
+			if istp {
+				if !runningStep || state.NextInProgress {
+					command = api.Continue
```

Comment:
Should be api.DirectionCongruentContinue or it won't work with recordings.

Reply:
Thanks for catching this. Done.